File: graph_algorithms/week1_graph_decomposition1/2_adding_exits_to_maze/connected_components.cpp
#include <iostream>
#include <vector>

using std::pair;
using std::vector;

// Marks all vertices reachable from u as visited.
void explore(vector<vector<int>> &adj, vector<bool> &visited, int u) {
  visited[u] = true;

  for (int v : adj[u])
    if (!visited[v])
      explore(adj, visited, v);
}

// Returns the number of connected components in the given undirected graph.
int number_of_components(vector<vector<int>> &adj) {
  int res = 0;

  vector<bool> visited = vector<bool>(adj.size(), false);
  for (int u = 0; u < adj.size(); u++)
    if (!visited[u]) {
      explore(adj, visited, u);
      res++;
    }

  return res;
}

int main() {
  size_t n, m;
  std::cin >> n >> m;

  vector<vector<int>> adj(n, vector<int>());
  for (size_t i = 0; i < m; i++) {
    int a, b;
    std::cin >> a >> b;
    adj[a - 1].push_back(b - 1);
    adj[b - 1].push_back(a - 1);
  }

  std::cout << number_of_components(adj) << std::endl;
}

File: graph_algorithms/week1_undirected_graphs/2_adding_exits_to_maze/connected_components.py
from typing import List


def explore(adj: List[List[int]], visited: List[bool], u: int) -> None:
    """Marks all vertices reachable from u as visited."""
    visited[u] = True

    for v in adj[u]:
        if not visited[v]:
            explore(adj, visited, v)


def number_of_components(adj: List[List[int]]) -> int:
    """Returns the number of connected components in the given undirected graph."""
    res = 0

    visited = [False] * len(adj)
    for u in range(len(adj)):
        if not visited[u]:
            explore(adj, visited, u)
            res += 1

    return res


if __name__ == "__main__":
    n, m = map(int, input().split())

    edges = [tuple(map(int, input().split())) for _ in range(m)]
    adj: List[List[int]] = [[] for _ in range(n)]
    for a, b in edges:
        adj[a - 1].append(b - 1)
        adj[b - 1].append(a - 1)

    print(number_of_components(adj))
